### PR TITLE
PartitionManager

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Implemented first stage of the class. Ready for initial merge.

### CODE REVIEW

**Comment by xezon** [comment]
Very nice. There are a few game bugs/exploits that involve this class, so will be nice to have it!

**Comment by duncanspumpkin** [comment]
It's a massive class unfortunately. I think what I'll do is hook up the ctor and implement only the very simplest of functions of it. All the more complex functions will be left as calls. Then the rest of it can be done in follow on PRs. This will allow any other classes that depend on it to be implemented whilst the meat of the functions are slowly worked on.

File: src/game/logic/system/partitionmanager.h [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        float *closestDistArg,
+        Coord3D *closestDistVecArg);
+    bool Try_Position(Coord3D const *center, float dist, float angle, FindPositionOptions const *options, Coord3D *result);
+    int Iterate_Cells_Along_Line(
+        Coord3D const &pos, Coord3D const &posOther, int (*proc)(PartitionCell *, void *), void *userData);
+    int Iterate_Cells_Breadth_First(Coord3D const *pos, int (*proc)(PartitionCell *, void *), void *userData);
+    void Remove_All_Dirty_Modules();
+    void Remove_From_Dirty_Modules(PartitionData *data);
+    void Shutdown();
+    void Clear_Pending_Undo_Shroud_Reveal_Queue();
```
Comment (jonwil) [inline src/game/logic/system/partitionmanager.h]: This one should be named Reset_Pending_Undo_Shroud_Reveal_Queue

**Review by jonwil — CHANGES_REQUESTED** [review]

**Review by jonwil — APPROVED** [review]

File: src/game/logic/system/partitionmanager.h [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    PartitionData *m_prev;
+    PartitionData *m_nextDirty;
+    PartitionData *m_prevDirty;
+    int32_t m_coiArrayCount;
+    int32_t m_coiInUseCount;
+    CellAndObjectIntersection *m_coiArray;
+    int32_t unk1;
+    int32_t m_dirtyFlag;
+    ShroudedStatusType m_shroudedness[16];
+    ShroudedStatusType m_ghostShroudedness[16];
```
Comment (tomsons26) [inline src/game/logic/system/partitionmanager.h]: seems this is m_shroudednessPrevious based on friend_setShroudednessPrevious setting this and end of getShroudedStatus setting it to the old value from m_shroudedness

File: src/game/logic/system/partitionmanager.h [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    bool m_unk4[16];
```
Comment (tomsons26) [inline src/game/logic/system/partitionmanager.h]: seems to be a is dirty or changed flag for shroudness?

File: src/game/logic/system/partitionmanager.h [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+public:
+    ~PartitionCell();
+
+    virtual void CRC_Snapshot(Xfer *xfer) override;
+    virtual void Xfer_Snapshot(Xfer *xfer) override;
+    virtual void Load_Post_Process() override{};
+
+private:
+    CellAndObjectIntersection *m_firstCoilInCell;
+    uint32_t unk1[16]; // some sort of 2 16bit structure
```
Comment (tomsons26) [inline src/game/logic/system/partitionmanager.h]: m_shroudStatus fetched by getShroudStatusForPlayer
its weird however, its accessed as ushort in addLooker and removeShrouder, union? bitfield?
hm or short m_shroudStatus[16][2]

File: src/game/logic/system/partitionmanager.h [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    float unk2;
```
Comment (tomsons26) [inline src/game/logic/system/partitionmanager.h]: PartitionManager::init sets these to value in range of -1000000.0 1000000.0, mac shows a _HUGE_DIST thats 1000000.0, possibly related

File: src/game/logic/system/partitionmanager.h (hunk 4 above)
Comment (tomsons26) [inline src/game/logic/system/partitionmanager.h]: 16 is a player count but not sure if we got a enum for that yet

File: src/game/logic/system/partitionmanager.h [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+{
+    VOT_THREAT,
+    VOT_VALUE,
+};
+
+enum ShroudedStatusType : int32_t
+{
+    SHROUDED_INVALID,
+    SHROUDED_NONE,
+    SHROUDED_PARTIAL,
```
Comment (tomsons26) [inline src/game/logic/system/partitionmanager.h]: invalid and none ok, not sure if i can accept partial or all

Comment (tomsons26) [reply src/game/logic/system/partitionmanager.h]: Yeah looking like 4 is all but very uncertain

**Comment by tomsons26** [comment]
are the args confirmed with ZH Mac? Not all symbols in rts.map match what ZH has as they changed the functions since the demo

File: src/game/logic/system/partitionmanager.h (hunk 2 above)
Comment (duncanspumpkin) [reply src/game/logic/system/partitionmanager.h]: Hmm, i was looking at the setters of this and it's always after a ghost check that was my thinking.

File: src/game/logic/system/partitionmanager.h (hunk 4 above)
Comment (duncanspumpkin) [reply src/game/logic/system/partitionmanager.h]: what is the convention for an unknown structure? Its definitily two uint16_t's unsure if its a struct or an array

File: src/game/logic/system/partitionmanager.h (hunk 3 above)
Comment (duncanspumpkin) [reply src/game/logic/system/partitionmanager.h]: I am unsure at this time so suggest keeping it as unk. It doesn't make too much sense for dirty when there is already a dirty flag.

File: src/game/logic/system/partitionmanager.h (hunk 4 above)
Comment (duncanspumpkin) [reply src/game/logic/system/partitionmanager.h]: I've gone for introducing a new struct called `ShrouderLooker` there is clearly a function that is add/remove looker and and/remove shrouder (possibly shroudGeneration)

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**